data/bookkeeping/txn_merkle.go

```diff
@@ -0,0 +1,98 @@
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// go-algorand is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with go-algorand.  If not, see <https://www.gnu.org/licenses/>.
+
+package bookkeeping
+
+import (
+	"fmt"
+
+	"github.com/algorand/go-algorand/crypto"
+	"github.com/algorand/go-algorand/crypto/merklearray"
+	"github.com/algorand/go-algorand/data/transactions"
+	"github.com/algorand/go-algorand/protocol"
+)
+
+// TxnMerkleTree returns a cryptographic commitment to the transactions in the
+// block, along with their ApplyData, as a Merkle tree.  This allows the
+// caller to either extract the root hash (for inclusion in the block
+// header), or to generate proofs of membership for transactions that are
+// in this block.
+func (block Block) TxnMerkleTree() (*merklearray.Tree, error) {
+	return merklearray.Build(&txnMerkleArray{block: block})
+}
+
+// txnMerkleArray is a representation of the transactions in this block,
+// along with their ApplyData, as an array for the merklearray package.
+type txnMerkleArray struct {
+	block Block
+}
+
+// Length implements the merklearray.Array interface.
+func (tma *txnMerkleArray) Length() uint64 {
+	return uint64(len(tma.block.Payset))
+}
+
+// Get implements the merklearray.Array interface.
+func (tma *txnMerkleArray) GetHash(pos uint64) (crypto.Digest, error) {
+	if pos >= uint64(len(tma.block.Payset)) {
+		return crypto.Digest{}, fmt.Errorf("txnMerkleArray.Get(%d) out of bounds %d", pos, len(tma.block.Payset))
+	}
+
+	var elem txnMerkleElem
+	elem.stib = tma.block.Payset[pos]
+
+	stxn, _, err := tma.block.DecodeSignedTxn(elem.stib)
+	if err != nil {
+		return crypto.Digest{}, err
+	}
+	elem.txn = stxn.Txn
+
+	return elem.Hash(), nil
+}
+
+// txnMerkleElem represents a leaf in the Merkle tree of all transactions
+// in a block.
+type txnMerkleElem struct {
+	txn  transactions.Transaction
+	stib transactions.SignedTxnInBlock
+}
+
+// ToBeHashed implements the crypto.Hashable interface.
+func (tme *txnMerkleElem) ToBeHashed() (protocol.HashID, []byte) {
+	// The leaf contains two hashes: the transaction ID (hash of the
+	// transaction itself), and the hash of the entire SignedTxnInBlock.
+	txid := tme.txn.ID()
+	stib := crypto.HashObj(&tme.stib)
+
+	var buf [2 * crypto.DigestSize]byte
+	copy(buf[:crypto.DigestSize], txid[:])
+	copy(buf[crypto.DigestSize:], stib[:])
+
+	return protocol.TxnMerkleLeaf, buf[:]
+}
+
+// Hash implements an optimized version of crypto.HashObj(tme).
+func (tme *txnMerkleElem) Hash() crypto.Digest {
```

Comment:
you're hashing both the `stib` as well as the `txid`, but eventually they both commit to the same underlying transaction. is there any benefit that I'm missing in creating a
hash(hash(txn)+hash(stxinblock)+nonce) vs just hash(stxinblock) ?

Reply:
The benefit of having the txid separately is that it simplifies verification of a transaction being committed, without requiring the verifier to decode the SignedTxnInBlock and determine if it's the same transaction as a particular txid it's looking for.

Reply:
yes, that's what I thought.. but that's not what you're doing - 
i.e. if you want to accomplish that, you need to have the actual merkle tree element contain the tx.ID() rather than a hash of the `txid` and `stib`. ( otherwise, you would have to have the hash(txib) in order to find it out ).

Reply:
Part of the proof is the hash of `stib`; the verifier doesn't have to do anything other than confirm that, this supplied `stib` value happens to be in the Merkle tree along with the txid it's expecting.

Reply:
right, I see that - but what it the added value here ? i.e. it sounds like we're forcing the user to have both of the values in all cases if he wants verify that it's in the Merkle tree. constructing the hash to be a "string" of the first hash followed by the second hash ( at leaf) would make much better "construct", as it would allow us to verify that the entry is there by comparing the bytes.

Reply:
That is exactly what's going on here.  `txid` and `stib` are `crypto.Digest` hashes of the transaction ID and the `SignedTxnInBlock` respectively.

Reply:
so, what I'm probably missing is the last line `return crypto.Hash(s)` - why do you hash the entire constructed array instead of just returning it ?

Reply:
Oh, I see what you are asking about.  This `Hash()` function returns the hash of the Merkle tree leaf, rather than the Merkle leaf itself.  (It is returned by `txnMerkleArray.GetHash()` above, and the `merklearray` code expects it to return the hash of the leaf.  I recently made this optimization/change in one of the earlier PRs leading up to this one.)

Reply:
given that the verifier is going to have just the hashes of the entire merkle tree ( and not the underlying data elements ), it would be a big saving if the verifier could prove that an element belongs to the tree just by looking on the hash value. this also consist with merkle proofs.

Reply:
another point ( unrelated to the previous one ):
do you think it would make sense to construct the array as
<hash(txid) + nonce + hash(stib)> instead of <nonce + hash(txid) +  hash(stib)> ?
I'm suggesting that so that we can perform log(n) lookups into the leafs instead of examining all of them.

Reply:
I'm probably missing what you're thinking about here; sorry about that.

The verifier just needs the hashes leading up to the root.  The prover can indeed prove that an element is in the tree by supplying the hash of the leaf (i.e., the value returned by `txnMerkleElem.Hash()`).  If the verifier knows the entire `SignedTxn` and its associated `ApplyData`, it can compute both the `stib` hash and the `txid` on its own.  If the verifier knows the `txid` it's looking for, but doesn't know (or doesn't want to compute) the entire `SignedTxnInBlock`, it can ask for the proof to also include the `stib` hash, so that it can concatenate its known `txid` with this supplied `stib` hash to verify that its `txid` lives in the tree.

We could construct two Merkle trees: one for just the transaction IDs, and one for just the `SignedTxnInBlock`s.  But it seems slightly more efficient to have a single Merkle tree where the leaves are the concatenation of the transaction ID and the `SignedTxnInBlock` hash.

Reply:
That's the thing - I don't think that we *need* to have two trees:
if the leaf hash would consist abit longer byte array ( let's say 66 bytes ), we can achieve both at the same time - 
the `prover` would provide a merkle proof that contains all the elements (sibling) from the root and up to the txn leaf.
The `verifier` see that the txn leaf "long hash" contain either the txid or the hash(stib), and can verify the path up to the root.
And when the verifier is asked to provide a proof for either txid or hash(stib), it can do so without rebuilding the SignedTxn but rather just iterate on the leaves and find the correct leaf in log(n) steps.

Reply:
I believe the design you describe is basically equivalent to the code in this PR.  The only difference, I think, is that you are describing this design in terms of having a Merkle tree of 2N nodes, and in this PR, I have this design implemented as using a `merklearray` Merkle tree of N nodes, where each node is effectively another 2-node subtree.

Reply:
I haven't thought about it in the terms of `N nodes merkle tree` vs `2N nodes merkle tree`, but you are correct in that my thinking would achieve that. The trigger for me was what seems to be a meaningless hashing - a hashing of previously hashed value, that only take away functionality.
I don't think that having your implementation as is would be a bad thing; In fact, from performance perspective, it's already slightly faster (7678093ns vs 6956778ns ). 
By skipping the hashing on the leaf node, I expect it to be ~10% faster, which would become meaningful if we were to increase the block size substentially.

Reply:
I don't believe there is any meaningless hashing going on here: this hash reduces `2*crypto.DigestSize` bytes to just `crypto.DigestSize` bytes, just like the hashes of internal Merkle tree nodes.

Reply:
no intend to insult, the meaningless is ( from my perspective ) is just because reducing the size isn't really our goal.

If I were to reduce the number of bytes, I would have taken the first 16 bytes from the first one and add the first 16 bytes from the second one. Given that the inputs are already "secure hashes" on their own, it wouldn't reduce the overall security of the solution, and provide look-ability into the hashes value themselves.

Another point is that not all the hashes in the merkle tree has to be of the same length. I know that it might seems abit "unorthodox" to do so, but if we can get better value of it, than it's probably worth considering.

Reply:
No insult taken -- I meant "meaningless" only in a technical sense.  :)

In some sense, I think reducing the size is pretty much the only goal of the Merkle tree.  Otherwise, the previous plan of hashing the entire payset is just as secure.

Indeed, having the same-size hashes in the Merkle tree at all layers simplifies the `merklearray` code, which is a nice benefit.

Reply:
Maybe as a way of understanding what you are hoping to improve: what operation do you think we could make more efficient?  I'm thinking there's three things going on with this Merkle tree: (1) online nodes generating the tree root when we build the block, (2) REST API nodes generating a proof for a specific transaction in a block, and (3) some other code that verifies a proof against a known txn root hash from a block header.

Reply:
I'm happy you've broken that like that; I think that all the above would benefit from that.
The online nodes generating the tree root, would not be required to make additional hashing call for the leaf elements.
The REST API nodes generating a proof for multiple transactions ids, would not need to calculate any hash(s). Instead, they would be able to perform a binary search on the leafs and find the leaf that matches the desired transaction. From there, the proof could be obtained by looking at the siblings.
And last, the verification of the leaf node becomes a simple byte array compare. This become meaningful when "exporting" the proof to another platform that isn't go-oriented where you don't want to have another "magic" number that would be "corresponding" to the transaction id on the Algorand platform. I can also see us supporting this on TEAL, but I'm not really sure in what form.

Reply:
For (1), I agree that the cost to generate the root commitment would be slightly lower if we were to concatenate the `txid` and `stib` in the leaf, rather than hash them together.

For (2), I believe the number of hashes needed to generate a proof is exactly the same (zero) once the tree is built.  After `merklearray` builds a tree, it retains all intermediate hashes, and generating a proof for any particular element just requires choosing the hashes along the path to the root.  This would be true both for generating a proof for one transaction or for multiple transactions.

For (3), I believe no byte comparison of leaves is needed in either case.  The procedure for verifying a proof, in both this PR's design and in your proposed alternative design, is to take the information we expect to be in the leaf node, and hash it (along with appropriate siblings) until we get a candidate root hash.  Then, we just compare the root hash we computed to the root hash we expect to see, from the block header.

One downside of the alternative design we're discussing here is that the proof becomes slightly longer: the first sibling is 2*64 bytes, instead of just 64 bytes long.

I think I'd prefer keeping the code as-is in this PR, since it leads to a simpler `merklearray` (supporting longer leaves would require changing that code), and I don't think there's a clear advantage to larger leaves.